src/AWS.Deploy.Orchestration/CustomRecipeLocator.cs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        /// <summary>
+        /// This method takes a root directory path and recursively searches all its sub-directories for custom recipe paths.
+        /// However, it ignores any recipe file located inside a "bin" folder.
+        /// </summary>
+        /// <param name="rootDirectoryPath">The absolute path of the root directory.</param>
+        /// <returns>A list of recipe definition paths.</returns>
+        private List<string> GetRecipePathsFromRootDirectory(string? rootDirectoryPath)
+        {
+            var recipePaths = new List<string>();
+            if (!string.IsNullOrEmpty(rootDirectoryPath) && _directoryManager.Exists(rootDirectoryPath))
+            {
+                foreach (var recipeFilePath in _directoryManager.GetFiles(rootDirectoryPath, "*.recipe", SearchOption.AllDirectories))
+                {
+                    if (recipeFilePath.Contains(_ignorePathSubstring))
+                        continue;
+                    recipePaths.Add(_directoryManager.GetDirectoryInfo(recipeFilePath).Parent.FullName);
+                }
+            }
+            return recipePaths;
+        }
+
+        /// <summary>
+        /// This method finds the root directory that is monitored by the same source control as the current directory.
+        /// </summary>
+        /// <param name="currentDirectoryPath">The absolute path of the current directory</param>
+        /// <returns>The source control root directory absolute path.</returns>
+        private async Task<string?> GetSourceControlRootDirectory(string currentDirectoryPath)
+        {
+            var possibleRootDirectoryPath = currentDirectoryPath;
+            while (currentDirectoryPath != null && await IsDirectoryUnderSourceControl(currentDirectoryPath))
+            {
+                possibleRootDirectoryPath = currentDirectoryPath;
+                currentDirectoryPath = _directoryManager.GetDirectoryInfo(currentDirectoryPath).Parent.FullName;
+            }
+            return possibleRootDirectoryPath;
+        }
+
+        /// <summary>
+        /// This method attempts to find the directory path which contains the solution file for the target application.
+        /// </summary>
+        /// <param name="targetApplicationDirectoryPath">The absolute directory path which contains the target application csproj or fsproj file.</param>
+        /// <returns>The absolute path of the directory which contains the solution file.</returns>
+        private string GetSolutionDirectory(string targetApplicationDirectoryPath)
+        {
+            var possiblesolutionDirectory = _directoryManager.GetDirectoryInfo(targetApplicationDirectoryPath).Parent.FullName;
+            var files = _directoryManager.GetFiles(possiblesolutionDirectory, "*.sln", SearchOption.TopDirectoryOnly);
+            if (files.Length == 1)
+            {
+                return possiblesolutionDirectory;
+            }
+            return string.Empty;
```

Comment:
This method should keep checking parent directories like how you are looking for source control root. Many project layouts have the solution file more then one folder up.

Reply:
Fixed by passing the `solutionDirectoryPath` which is resolved using the `ProjectDefinition` member of the `OrchestratorSession`